File: App/Resources/AlgorithmDetails/blockinsertionsort/blockinsertionsort.c
#include <stdio.h>
#include <stdlib.h>

int array[16] = {0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56};

void swap(int arr[], int a, int b) {
  int t = arr[a];
  arr[a] = arr[b];
  arr[b] = t;
}

void printList(int items[], int size) {
  for (int i = 0; i < size; i++) {
    if (i == 0) {
      printf("[%d, ", items[i]);
    } else if (i != size - 1) {
      printf("%d, ", items[i]);
    } else {
      printf("%d]", items[i]);
    }
  }
}

void multiSwap(int arr[], int a, int b, int count) {
  for (int i = 0; i < count; i++)
    swap(arr, a + i, b + i);
}

void rotate(int arr[], int pos, int lenA, int lenB) {
  while (lenA != 0 && lenB != 0) {
    if (lenA <= lenB) {
      multiSwap(arr, pos, pos + lenA, lenA);
      pos += lenA;
      lenB -= lenA;
    } else {
      multiSwap(arr, pos + (lenA - lenB), pos + lenA, lenB);
      lenA -= lenB;
    }
  }
}

int binSearch(int arr[], int pos, int len, int keyPos, int isLeft) {
  int left = 0, right = len;
  while (left < right) {
    int mid = left + (right - left) / 2;
    int cond = isLeft ? (arr[pos + mid] < arr[keyPos])
                      : (arr[pos + mid] <= arr[keyPos]);
    if (cond)
      left = mid + 1;
    else
      right = mid;
  }
  return left;
}

void mergeWithoutBuffer(int arr[], int pos, int len1, int len2) {
  if (len1 < len2) {
    while (len1 != 0) {
      int loc = binSearch(arr, pos + len1, len2, pos, 1);
      if (loc != 0) { rotate(arr, pos, len1, loc); pos += loc; len2 -= loc; }
      if (len2 == 0) break;
      do { pos++; len1--; } while (len1 != 0 && arr[pos] <= arr[pos + len1]);
    }
  } else {
    while (len2 != 0) {
      int loc = binSearch(arr, pos, len1, pos + len1 + len2 - 1, 0);
      if (loc != len1) { rotate(arr, pos + loc, len1 - loc, len2); len1 = loc; }
      if (len1 == 0) break;
      do { len2--; } while (len2 != 0 && arr[pos + len1 - 1] <= arr[pos + len1 + len2 - 1]);
    }
  }
}

int findRun(int arr[], int a, int b) {
  int i = a + 1;
  if (i == b)
    return i;
  if (arr[i - 1] > arr[i]) {
    i++;
    while (i < b && arr[i - 1] > arr[i])
      i++;
    int lo = a, hi = i - 1;
    while (lo < hi) {
      swap(arr, lo, hi);
      lo++;
      hi--;
    }
  } else {
    i++;
    while (i < b && arr[i - 1] <= arr[i])
      i++;
  }
  return i;
}

void insert1(int arr[], int a, int l) {
  int tmp = arr[l];
  l--;
  while (l >= a && arr[l] > tmp) {
    arr[l + 1] = arr[l];
    l--;
  }
  arr[l + 1] = tmp;
}

void insert2(int arr[], int a, int l, int r) {
  int tmpL = arr[l];
  int tmpR = arr[r];
  l--;
  while (l >= a && arr[l] > tmpR) {
    arr[l + 2] = arr[l];
    l--;
  }
  arr[l + 2] = tmpR;
  while (l >= a && arr[l] > tmpL) {
    arr[l + 1] = arr[l];
    l--;
  }
  arr[l + 1] = tmpL;
}

void sort(int arr[], int n) {
  if (n < 2) return;
  int i = findRun(arr, 0, n);
  while (i < n) {
    int j = findRun(arr, i, n);
    int len = j - i;
    if (len == 1)
      insert1(arr, 0, i);
    else if (len == 2)
      insert2(arr, 0, i, i + 1);
    else
      mergeWithoutBuffer(arr, 0, i, len);
    i = j;
  }
}

int main(int argc, char *argv[]) {
  int size = sizeof(array) / sizeof(array[0]);
  sort(array, size);
  printList(array, size);
  return 0;
}

File: App/Resources/AlgorithmDetails/blockinsertionsort/blockinsertionsort.py
def multi_swap(arr, a, b, count):
    for i in range(count):
        arr[a + i], arr[b + i] = arr[b + i], arr[a + i]


def rotate(arr, pos, len_a, len_b):
    while len_a != 0 and len_b != 0:
        if len_a <= len_b:
            multi_swap(arr, pos, pos + len_a, len_a)
            pos += len_a
            len_b -= len_a
        else:
            multi_swap(arr, pos + (len_a - len_b), pos + len_a, len_b)
            len_a -= len_b


def bin_search(arr, pos, length, key_pos, is_left):
    left, right = 0, length
    while left < right:
        mid = left + (right - left) // 2
        cond = (
            arr[pos + mid] < arr[key_pos] if is_left else arr[pos + mid] <= arr[key_pos]
        )
        if cond:
            left = mid + 1
        else:
            right = mid
    return left


def merge_without_buffer(arr, pos, len1, len2):
    if len1 < len2:
        while len1:
            loc = bin_search(arr, pos + len1, len2, pos, True)
            if loc:
                rotate(arr, pos, len1, loc)
                pos += loc
                len2 -= loc
            if not len2:
                break
            while True:
                pos += 1
                len1 -= 1
                if not len1 or arr[pos] > arr[pos + len1]:
                    break
    else:
        while len2:
            loc = bin_search(arr, pos, len1, pos + len1 + len2 - 1, False)
            if loc != len1:
                rotate(arr, pos + loc, len1 - loc, len2)
                len1 = loc
            if not len1:
                break
            while True:
                len2 -= 1
                if not len2 or arr[pos + len1 - 1] > arr[pos + len1 + len2 - 1]:
                    break


def find_run(arr, a, b):
    i = a + 1
    if i == b:
        return i
    if arr[i - 1] > arr[i]:
        i += 1
        while i < b and arr[i - 1] > arr[i]:
            i += 1
        lo, hi = a, i - 1
        while lo < hi:
            arr[lo], arr[hi] = arr[hi], arr[lo]
            lo += 1
            hi -= 1
    else:
        i += 1
        while i < b and arr[i - 1] <= arr[i]:
            i += 1
    return i


def insert1(arr, a, l):
    tmp = arr[l]
    l -= 1
    while l >= a and arr[l] > tmp:
        arr[l + 1] = arr[l]
        l -= 1
    arr[l + 1] = tmp


def insert2(arr, a, l, r):
    tmp_l = arr[l]
    tmp_r = arr[r]
    l -= 1
    while l >= a and arr[l] > tmp_r:
        arr[l + 2] = arr[l]
        l -= 1
    arr[l + 2] = tmp_r
    while l >= a and arr[l] > tmp_l:
        arr[l + 1] = arr[l]
        l -= 1
    arr[l + 1] = tmp_l


def sort(arr):
    n = len(arr)
    if n < 2:
        return
    i = find_run(arr, 0, n)
    while i < n:
        j = find_run(arr, i, n)
        length = j - i
        if length == 1:
            insert1(arr, 0, i)
        elif length == 2:
            insert2(arr, 0, i, i + 1)
        else:
            merge_without_buffer(arr, 0, i, length)
        i = j


if __name__ == "__main__":
    array = [0, 39, 21, 62, 91, 77, 14, 23, 90, 69, 51, 81, 68, 83, 32, 56]
    sort(array)
    print(array)
